src/main/java/net/minecraftforge/fluids/FluidInteractionRegistry.java

```diff
@@ -0,0 +1,179 @@
+package net.minecraftforge.fluids;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.LiquidBlock;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.material.FluidState;
+import net.minecraftforge.common.ForgeMod;
+import net.minecraftforge.event.ForgeEventFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * A registry which defines the interactions a source fluid can have with its
+ * surroundings. Each possible flow direction is checked for all interactions with
+ * the source.
+ */
+public final class FluidInteractionRegistry
+{
+    private static final Map<FluidType, List<InteractionInformation>> INTERACTIONS = new HashMap<>();
+
+    /**
+     * Adds an interaction between a source and its surroundings.
+     *
+     * @param source the source of the interaction, this will be replaced if the interaction occurs
+     * @param interaction the interaction data to check and perform
+     */
+    public static synchronized void addInteraction(FluidType source, InteractionInformation interaction)
+    {
+        INTERACTIONS.computeIfAbsent(source, s -> new ArrayList<>()).add(interaction);
+    }
+
+    /**
+     * Performs all potential fluid interactions at a given position.
+     *
+     * <p>Note: Only the first interaction check that succeeds will occur.
+     *
+     * @param level the level the interactions take place in
+     * @param pos the position of the source fluid
+     * @return {@code true} if an interaction took place, {@code false} otherwise
+     */
+    public static boolean canInteract(Level level, BlockPos pos)
+    {
+        FluidState state = level.getFluidState(pos);
+        for (Direction direction : LiquidBlock.POSSIBLE_FLOW_DIRECTIONS)
+        {
+            BlockPos relativePos = pos.relative(direction.getOpposite());
+            List<InteractionInformation> interactions = INTERACTIONS.getOrDefault(state.getFluidType(), Collections.emptyList());
+            for (InteractionInformation interaction : interactions)
+            {
+                if (interaction.predicate().test(level, pos, relativePos, state))
+                {
+                    interaction.interaction().interact(level, pos, relativePos, state);
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    static
+    {
```

Comment:
Is water + lava -> stone just missing here or is there some reason it doesn't fall under this interaction system